database/database_clouding.py

```diff
@@ -0,0 +1,132 @@
+from google.cloud import spanner
+
+# Instantiates a client
+spanner_client = spanner.Client()
+
+# Your Cloud Spanner instance ID.
+instance = spanner_client.instance('salk-demo')
+
+# Get a Cloud Spanner database by ID.
+database = instance.database("users")
+
+# Dummy data
+def add_vm_to_database(hostname: str): 
+    """
+    Adds a hostname of a Virtual Machine to the database with a default pin and command (None)
+
+    Parameters:
+    hostname (str): The hostname of the VM to be added to the database
+    """
+    with database.batch() as batch:
+        # Checks if there is already a VM with the given hostname
+        with database.snapshot() as snapshot:
+            result = snapshot.execute_sql(
+                "SElECT * FROM Users WHERE HostName = @hostname",
+                params={'hostname': hostname},
+                param_types={'hostname': spanner.param_types.STRING}
+            )
+            for row in result:
+                print("The hostname", hostname, "already exists in the database")
+                return
+
+        batch.insert(
+            table="Users",
+            columns=("HostName", "Pin", "CrdCmd", "UserEmail"),
+            values=[(hostname, None, None, None)]
+        )
+    print(hostname, "added VM to database")
+
+def add_cmd_and_pin(hostname: str, pin: str, cmd: str):
+    """
+    Adds a command and pin to the given hostname in the database when a user connects to the VM with the corresponding hostname
+
+    Parameters:
+    hostname (str): The hostname of the VM to be added to the database
+    pin (str): The pin that user sets to be added to the database
+    cmd (str): The command to be added to the database
+    """
+    # TODO: Add error handling for if the hostname does not exist in the database
+
+    # Checks if the hostname already has a pin or command)
+
+    if (get_pin_and_crd(hostname) != (None, None)):
+        raise ValueError("The hostname", hostname, "already has a pin or command")
```

Comment:
Let's have an optional override on this (say if we want to remove a VM assignment - clear the values for pin and command)
```suggestion
    if avoid_overwrite and (get_pin_and_crd(hostname) != (None, None)):
        raise ValueError("The hostname", hostname, "already has a pin or command")
```